File: plugin/vscode/src/test/service/mergeIssues.test.ts
import * as assert from "node:assert/strict";
import { Issue } from "../../models/Issue";
import { mergeIssues } from "../../service/mergeIssues";

suite("mergeIssues", () => {
  test("returns new issues when no global error", () => {
    const merged = mergeIssues(
      [{ line: 1, message: "old", exit: 1 }],
      [{ line: 2, message: "new", exit: 1 }],
    );

    assert.deepEqual(merged, [{ line: 2, message: "new", exit: 1 }]);
  });

  test("keeps prior line issues when global error", () => {
    const merged = mergeIssues(
      [
        { line: 1, message: "stale line", exit: 1 },
        { line: null, message: "old global", exit: 2 },
      ],
      [
        { line: null, message: "fresh global", exit: 2 },
        { line: 9, message: "ignored line from cli", exit: 1 },
      ],
    );

    assert.deepEqual(merged, [
      { line: 1, message: "stale line", exit: 1 },
      { line: null, message: "fresh global", exit: 2 },
    ]);
  });

  test("treats exit 2 with line as not global error", () => {
    const issues: Issue[] = [{ line: 3, message: "line error", exit: 2 }];
    assert.deepEqual(mergeIssues([], issues), issues);
  });

  test("global error drops line issues when cache empty", () => {
    const merged = mergeIssues(undefined, [
      { line: null, message: "global", exit: 2 },
      { line: 5, message: "line", exit: 1 },
    ]);

    assert.deepEqual(merged, [{ line: null, message: "global", exit: 2 }]);
  });
});

File: plugin/intellij/src/test/kotlin/com/github/jshwi/docsig/service/MergeIssuesTest.kt
package com.github.jshwi.docsig.service

import com.github.jshwi.docsig.models.Issue
import org.junit.jupiter.api.Test
import kotlin.test.assertEquals

/**
 * The vscode extension and the neovim plugin merge the same way, so
 * these cases mirror theirs.
 */
class MergeIssuesTest {
    @Test
    fun `a clean run replaces the cache`() {
        val previous = listOf(Issue(1, "old", 1))
        val issues = listOf(Issue(2, "new", 1))

        assertEquals(issues, mergeIssues(previous, issues))
    }

    @Test
    fun `a global error keeps previous line issues`() {
        val previous = listOf(Issue(1, "old", 1), Issue(null, "stale", 2))
        val issues = listOf(Issue(null, "cli blew up", 2))

        assertEquals(
            listOf(Issue(1, "old", 1), Issue(null, "cli blew up", 2)),
            mergeIssues(previous, issues),
        )
    }

    @Test
    fun `a global error with no cache yields the new issues`() {
        val issues = listOf(Issue(null, "cli blew up", 2))

        assertEquals(issues, mergeIssues(null, issues))
    }

    @Test
    fun `a line level error is not treated as global`() {
        val previous = listOf(Issue(1, "old", 1))
        val issues = listOf(Issue(3, "syntax error", 2))

        assertEquals(issues, mergeIssues(previous, issues))
    }

    @Test
    fun `an empty run clears the cache`() {
        val previous = listOf(Issue(1, "old", 1))

        assertEquals(emptyList(), mergeIssues(previous, emptyList()))
    }
}
